File: ISchemm.UTF32.Strings/String32.fs
namespace ISchemm.UTF32.Strings

open ISchemm.UTF32
open ISchemm.UTF32.Arrays
open System.Collections
open System.Collections.Generic

type String32 = {
    List: Char32 list
} with
    member this.Substring (startIndex: int) =
        { List = this.List.[startIndex..] }

    member this.Substring (startIndex: int, length: int) =
        { List = this.List.[startIndex..startIndex+length-1] }

    static member FromEnumerable (src: seq<Char32>) =
        { List = List.ofSeq src }

    static member FromString (str: string) =
        str
        |> Char32Array.FromString
        |> String32.FromEnumerable

    override this.ToString () =
        Array.ofList this.List
        |> Char32Array.GetString

    interface seq<Char32> with
        member this.GetEnumerator(): IEnumerator =
            (this.List :> System.Collections.IEnumerable).GetEnumerator()
        member this.GetEnumerator(): IEnumerator<Char32> = 
            (this.List :> seq<Char32>).GetEnumerator()

    static member op_Equality (left: String32, right: String32) =
        left = right

    static member op_Inequality (left: String32, right: String32) =
        left <> right

    static member op_LessThanOrEqual (left: String32, right: String32) =
        left <= right

    static member op_GreaterThanOrEqual (left: String32, right: String32) =
        left >= right

    static member op_LessThan (left: String32, right: String32) =
        left < right

    static member op_GreaterThan (left: String32, right: String32) =
        left > right

File: ISchemm.UTF32.Strings/String32.cs
using ISchemm.UTF32.Arrays;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace ISchemm.UTF32.Strings
{
    public struct String32 : IEnumerable<Char32>, IEquatable<String32>
    {
        private readonly Char32[] _array;

        public String32(Char32[] array)
        {
            _array = array ?? throw new ArgumentNullException(nameof(array));
        }

        public override int GetHashCode()
        {
            int hash = 17;
            foreach (Char32 element in _array)
            {
                hash = hash * 31 + element.GetHashCode();
            }
            return hash;
        }

        public override bool Equals(object obj)
        {
            return obj is String32 str && Equals(str);
        }

        public bool Equals(String32 other)
        {
            return _array.SequenceEqual(other._array);
        }

        public String32 Substring(int startIndex)
        {
            return new String32(_array.Skip(startIndex).ToArray());
        }

        public String32 Substring(int startIndex, int length)
        {
            return new String32(_array.Skip(startIndex).Take(length).ToArray());
        }

        public static String32 FromEnumerable(IEnumerable<Char32> src)
        {
            return new String32(src.ToArray());
        }

        public static String32 FromString(string str)
        {
            return FromEnumerable(Char32Array.FromString(str));
        }

        public override string ToString()
        {
            return Char32Array.GetString(_array);
        }

        public IEnumerator<Char32> GetEnumerator()
        {
            return ((IEnumerable<Char32>)_array).GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return _array.GetEnumerator();
        }
    }
}
